src/box/memtx_tree.cc

```diff
@@ -932,25 +975,59 @@ tree_iterator_start(struct iterator *iterator, struct tuple **ret)
 	struct tuple *successor = initial_elem ? initial_elem->tuple : NULL;
 
 	struct memtx_tree_data<USE_HINT> *res = initial_elem;
-	if (iterator_type_is_reverse(type))
+
+	/*
+	 * If the iterator type is not reverse and we don't skip tuples, the
+	 * initial_elem is the result of the first iteration step. Otherwise
+	 * the lookup function performs an extra step back or tuple skip, so
+	 * we need to actualize the current element after it.
+	 */
+	if (iterator_type_is_reverse(type) || it->skip != 0)
 		res = memtx_tree_iterator_get_elem(tree, &it->tree_iterator);
 
+	/*
+	 * It we have skipped raw tuples, some of them may be invisible to the
+	 * current transaction, let's skip further in this case.
+	 */
+	struct txn *txn = in_txn();
+	if (it->skip != 0 && res != NULL) {
+		size_t need_skip_more_visible =
+			memtx_tx_index_invisible_count_matching_until(
+				in_txn(), space, index_base, type,
+				start_data.key, start_data.part_count,
+				res->tuple, res->hint);
+
+		/* Skip more visible tuples. */
+		memtx_tree_iterator_t<USE_HINT> *iterator = &it->tree_iterator;
+		while (need_skip_more_visible && res != NULL) {
+			if (memtx_tx_tuple_is_visible(txn, index_base,
+						      res->tuple))
+				need_skip_more_visible--;
+			if (iterator_type_is_reverse(type))
+				memtx_tree_iterator_prev(tree, iterator);
+			else
+				memtx_tree_iterator_next(tree, iterator);
+			res = memtx_tree_iterator_get_elem(tree, iterator);
+		}
+	}
+
+	bool is_eq = type == ITER_EQ || type == ITER_REQ;
+
 	/* If we skip tuple, flag equals is not actual - need to refresh it. */
-	if (it->after_data.key != NULL && res != NULL &&
-	    (type == ITER_EQ || type == ITER_REQ)) {
+	if (((it->after_data.key != NULL && is_eq) || it->skip != 0) &&
+	    res != NULL) {
```

Comment:
Nit: you do some microrefactoring across this patch, like replacing `(type == ITER_EQ || type == ITER_REQ)` with `is_eq `. It would be a little simpler to review if you would try to factor them out into separate patches.

Reply:
Done.